src/features/bundlr/BundlrInput.tsx

```diff
@@ -4,35 +4,89 @@ import {
   NumberInputStepper,
   NumberIncrementStepper,
   NumberDecrementStepper,
-  NumberInputProps,
+  InputGroup,
+  InputRightElement,
 } from '@chakra-ui/react';
+import { ethers } from 'ethers';
+import { uploadPriceDecimals } from 'lib/constants';
+import { useState } from 'react';
 
-export function BundlrInput(props: NumberInputProps) {
-  return (
-    <NumberInput
-      step={0.01}
-      mt={3}
-      w="100%"
-      min={0}
-      color="whiteAlpha.700"
-      allowMouseWheel
-      {...props}
-    >
-      <NumberInputField
-        borderColor="whiteAlpha.700"
-        color="whiteAlpha.800"
-        h="48px"
-      />
-      <NumberInputStepper>
-        <NumberIncrementStepper
-          color="whiteAlpha.700"
-          borderColor="whiteAlpha.700"
-        />
-        <NumberDecrementStepper
-          color="whiteAlpha.700"
+const inputLimit = 46;
+
+export function useBundlrInput(args: { initialAmount?: number; inputHeight?: string } = {}) {
+  const { initialAmount, inputHeight: heightArg } = args;
+
+  const [inputAmount, setInputAmount] = useState(
+    `${initialAmount?.toFixed(uploadPriceDecimals) ?? ''}`
+  );
+  const [isValidInput, setIsValidInput] = useState(false);
+
+  const handleInputChange = (valueAsString: string) => {
+    setIsValidInput(false);
+
+    if (!valueAsString) {
+      setInputAmount('');
+      return;
+    }
+
+    valueAsString = valueAsString.slice(0, inputLimit);
+    const valueAsFloat = Number.parseFloat(valueAsString);
+
+    setIsValidInput(!Number.isNaN(valueAsFloat) && valueAsFloat > 0);
+
+    try {
+      ethers.utils.parseEther(valueAsString);
+    } catch (_) {
+      setIsValidInput(false);
+    }
+
+    setInputAmount(valueAsString);
+  };
+
+  const inputHeight = heightArg ?? '48px';
+
+  const BundlrInput = (
```

Comment:
`BundlrInput` should be a function that returns the JSX. 
```
const BundlrInput = () => {
  return <>...</>
}
```

Reply:
Then you should be able to use the component like `<BundlrInput />` instead of `{BundlrInput}`.

Reply:
I would also consider passing in the `inputHeight` prop into the component instead of the hook if possible. Something like this: 
```
const BundlrInput = ({inputHeight}) => {
  return <Flex h={inputHeight ?? "48px"}>...</Flex>
}
```

Reply:
This is the pattern I went with at first, but it was causing the input to lose focus whenever anything is typed in it. I'm guessing it's something to do with it being recreated in the `useBundlrInput` hook.

As for my going with a component-returning hook, it's mainly so I could abstract input change away from the consumer, and only expose the processed input, while making sure updates to it would be seen by the consumer.

There may other advantages that you saw that I didn't think of -- would love to know if any. 

Also if the goal above can easily be achieved using components only, that'd be super to know. Orrr if you know of a fix for the lost-focus bug, then we can keep the hook and return a JSX-returning function. 

Otherwise we'll have to break the recommended pattern and leave it as is.

Reply:
Hum I think I've figured it out lol. I'll update this hook to be a component instead. Looks like I can still keep the abstraction I want. The focus bug doesn't seem to be present this way too.